velox/exec/TopNRowNumber.h

```diff
@@ -0,0 +1,213 @@
+        : rows{{comparator}, StlAllocator<char*>(allocator)} {};
+  };
+
+  /// A synthetic Aggregate passed to HashTable and its RowContainer to
+  /// associate a 'TopRows' struct with each partition.
+  ///
+  /// TODO Refactor RowContainer and HashTable to not require an 'Aggregate'
+  /// instance.
+  class TopNAggregate : public Aggregate {
+   public:
+    TopNAggregate(RowComparator comparator)
+        : Aggregate(BIGINT()), comparator_{comparator} {}
+
+    int32_t accumulatorFixedWidthSize() const override {
+      return sizeof(TopRows);
+    }
+
+    TopRows& topRows(char* group) {
+      return *reinterpret_cast<TopRows*>(group + offset_);
+    }
+
+    void initializeNewGroups(
+        char** groups,
+        folly::Range<const vector_size_t*> indices) override {
+      for (auto index : indices) {
+        new (groups[index] + offset_) TopRows(allocator_, comparator_);
+      }
+    }
+
+    void addRawInput(
+        char** groups,
+        const SelectivityVector& rows,
+        const std::vector<VectorPtr>& args,
+        bool mayPushdown) override {
+      VELOX_UNREACHABLE();
+    }
+
+    void addIntermediateResults(
+        char** groups,
+        const SelectivityVector& rows,
+        const std::vector<VectorPtr>& args,
+        bool mayPushdown) override {
+      VELOX_UNREACHABLE();
+    }
+
+    void addSingleGroupRawInput(
+        char* group,
+        const SelectivityVector& rows,
+        const std::vector<VectorPtr>& args,
+        bool mayPushdown) override {
+      VELOX_UNREACHABLE();
+    }
+
+    void addSingleGroupIntermediateResults(
+        char* group,
+        const SelectivityVector& rows,
+        const std::vector<VectorPtr>& args,
+        bool mayPushdown) override {
+      VELOX_UNREACHABLE();
+    }
+
+    void extractValues(char** groups, int32_t numGroups, VectorPtr* result)
+        override {
+      VELOX_UNREACHABLE();
+    }
+
+    void extractAccumulators(
+        char** groups,
+        int32_t numGroups,
+        VectorPtr* result) override {
+      VELOX_UNREACHABLE();
+    }
+
+   private:
+    RowComparator comparator_;
+  };
+
+  /// Adds input row to a partition or discards the row.
+  void processInputRow(
+      const RowVectorPtr& input,
+      vector_size_t index,
+      TopRows& partition);
+
+  /// Returns next partition to add to output or nullptr if there are no
+  /// partitions left.
+  TopRows* nextPartition();
+
+  /// Returns partition that was partially added to the previous output batch.
+  TopRows& currentPartition();
+
+  /// Appends partition rows to outputRows_ and optionally populates row
+  /// numbers.
+  void appendPartitionRows(
+      TopRows& partition,
+      vector_size_t start,
+      vector_size_t size,
+      vector_size_t outputOffset,
+      FlatVector<int64_t>* rowNumbers);
+
+  const int32_t limit_;
+  const bool generateRowNumber_;
+  const RowTypePtr inputType_;
+
+  /// Hash table to keep track of partitions. Not used if there are no
```

Comment:
Are /// required here ? And for the rest of the variables in this class since they needn't be externally documented.